draft-ietf-tls-tls13.md

```diff
@@ -2419,7 +2445,44 @@ the existing behavior for SignedCertificateTimestamps {{?RFC6962}},
 and is more easily extensible in the handshake state machine.
 
 
-## Server Parameters Messages
+### PSK Binder
+
+%%% Authentication Messages
+       struct {
+           opaque binding_data[Hash.length];
+       } PskBinder;
+
+The "psk_binder" extension is used to carry a binding between the PSK
+and the current handshake. as well as between the session where the
+PSK was established and the session where it was used. This binding
+transitively includes the original handshake transcript, because that
+transcript is digested into the values which produce the Resumption
+Master Secret. This requires that both the KDF used to produce the RMS
+and the MAC used to compute the psk_binder be collision
+resistant. These are properties of HKDF and HMAC respectively when
+used with collision resistant hash functions and producing output of
+at least 256 bits.  Any future replacement of these functions MUST
+also provide collision resistance.
+
+The "psk_binder" extension MUST be sent if the client offers a
+"pre_shared_key" extension and MUST NOT be sent otherwise. The Hash
+used in the computation is that associated with the indicated PSK.
+Servers MUST NOT send this extension. This extension MUST be the
+last extension in the ClientHello (this facilitates implementation
+as described below). Servers MUST check that it is the last
+extension and otherwise fail the handshake with an "illegal_parameter"
+alert.
+
+PskBinder.binding_data is computed using the same way as the
+Finished.verify data value, using the PSK as the base key and the
+portion of the handshake transcript up to but not including
+PskBinder.binding_data as the Handshake Context. If the handshake
+includes a HelloRetryRequest, the initial ClientHello and
+HelloRetryRequest are included in the transcript.  The actual
+ClientHello, complete with the full "psk_binder" extension is included
+in all other handshake hash computations.
```

Comment:
This needs to be even more clear, because the risk of getting this wrong is very high.

You need to make it clear that this includes the ClientHello that the "psk_binder" extension is carried.  And that the ClientHello length fields are constructed as though the "psk_binder" extension were present.  And that the entire extension including identifier and length is omitted from the hash.

I think that an example, or illustration would help.  Making the same point several different ways will avoid mistakes and misinterpretations.

For the same reason, you should include a formula for the calculation:

    HMAC(Hash, PSK, [ClientHello + HelloRetryRequest +] Truncated ClientHello)